pkg/volume/glusterfs/glusterfs.go

```diff
@@ -420,6 +443,122 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
 	return d.plugin.host.GetPodVolumeDir(d.glusterfsMounter.glusterfs.pod.UID, strings.EscapeQualifiedNameForDisk(name), d.glusterfsMounter.glusterfs.volName)
 }
 
+//
+// Traverse the PVs, fetching all the GIDs from those
+// in a given storage class, and mark them in the table.
+//
+func (p *glusterfsPlugin) collectGids(className string, gidTable *MinMaxAllocator) error {
+	pvList, err := p.host.GetKubeClient().Core().PersistentVolumes().List(v1.ListOptions{LabelSelector: labels.Everything().String()})
+	if err != nil {
+		glog.Errorf("glusterfs: failed to get existing persistent volumes")
+		return err
+	}
+
+	for _, pv := range pvList.Items {
+		if storageutil.GetVolumeStorageClass(&pv) != className {
+			continue
+		}
+
+		gidStr, ok := pv.Annotations[volumehelper.VolumeGidAnnotationKey]
+
+		if !ok {
+			glog.Errorf("glusterfs: no gid found in pv")
+			continue
+		}
+
+		gid, err := convertGid(gidStr)
+		if err != nil {
+			glog.Errorf("glusterfs: failed to convert gid %v", gidStr)
+			return err
+		}
+
+		_, err = gidTable.Allocate(int(gid))
+		if err == ErrConflict {
+			glog.Errorf("glusterfs: warnng: gid %v already allocated", gid)
+		} else if err != nil {
+			glog.Errorf("glusterfs: failed to store gid: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+//
+// Return the gid table for a storage class.
+// - If this is the first time, fill it with all the gids
+//   used in PVs of this storage class by traversing the PVs.
+// - Adapt the range of the table to the current range of the SC.
+//
+func (p *glusterfsPlugin) getGidTable(className string, min uint32, max uint32) (*MinMaxAllocator, error) {
+	var err error
+	p.gidTableLock.Lock()
+	gidTable, ok := p.gidTable[className]
+	p.gidTableLock.Unlock()
+
+	if ok {
+		err = gidTable.SetRange(int(min), int(max))
+		if err != nil {
+			return nil, err
+		}
+
+		return gidTable, nil
+	}
+
+	// create a new table and fill it
+	newGidTable, err := NewMinMaxAllocator(0, absoluteGidMax)
+	if err != nil {
+		return nil, err
+	}
+
+	// collect gids with the full range
+	err = p.collectGids(className, newGidTable)
+	if err != nil {
+		return nil, err
+	}
+
+	// and only reduce the range afterwards
+	err = newGidTable.SetRange(int(min), int(max))
+	if err != nil {
+		return nil, err
+	}
+
+	// if in the meantime a table appeared, use it
+
+	p.gidTableLock.Lock()
+	defer p.gidTableLock.Unlock()
+
+	gidTable, ok = p.gidTable[className]
+	if ok {
+		err = gidTable.SetRange(int(min), int(max))
+		if err != nil {
+			return nil, err
+		}
+
+		return gidTable, nil
+	}
+
+	p.gidTable[className] = newGidTable
+
+	return newGidTable, nil
+}
+
+func (d *glusterfsVolumeDeleter) getGid() (uint32, error) {
+	gidStr, ok := d.spec.Annotations[volumehelper.VolumeGidAnnotationKey]
+
+	if !ok {
+		return 0, ErrNotFound
+	}
+
+	gid, err := convertGid(gidStr)
```

Comment:
if the annotation is malformed, the code currently logs about three times

convertGid(), line 408:
glog.Errorf("glusterfs: failed to parse gid %v ", inputGid)

getGid(), line 555:
glog.Errorf("glusterfs: failed to parse gid %v", gid)

Delete(), line 582:
glog.Errorf("glusterfs: failed to parse the gid")

instead, convertGid() should return an error like this, without logging:
```
return 0, fmt.Errorf("glusterfs: failed to parse gid %v: %v", inputGid, err)
```

getGid() can then just do this, also without logging:
```
gid, err := convertGid(gidStr)
return gid, true, err
```

and Delete() can log the result without wrapping the error:
```
if gid, exists, err := d.getGid(); err != nil {
  // being unable to get the gid from the PV should not stop deletion of it
  glog.Error(err)
} else if exists {
  gidTable, err := d.plugin.getGidTable(class.Name, cfg.gidMin, cfg.gidMax)
  if err != nil {
    return err
  }
  if err := gidTable.Release(int(gid)); if err != nil {
    return err
  }
}
```